pywr/parameters/rbf.py

```diff
@@ -0,0 +1,331 @@
+                values.extend(x.values)
+
+        for parameter, x in self.parameters.items():
+            if x.is_variable:
+                values.extend(x.values)
+
+        if self.days_of_year is not None:
+            if self.days_of_year.is_variable:
+                values.extend(self.days_of_year.values)
+
+        if self.y.is_variable:
+            values.extend(self.y.values)
+
+        return np.array(values)
+
+    def get_double_upper_bounds(self):
+        """Returns an array of the upper bounds of the variables."""
+        values = []
+
+        for node, x in self.nodes.items():
+            if x.is_variable:
+                values.extend(x.get_upper_bounds())
+
+        for parameter, x in self.parameters.items():
+            if x.is_variable:
+                values.extend(x.get_upper_bounds())
+
+        if self.days_of_year is not None:
+            if self.days_of_year.is_variable:
+                values.extend(self.days_of_year.get_upper_bounds())
+
+        if self.y.is_variable:
+            values.extend(self.y.get_upper_bounds())
+
+        return np.array(values)
+    
+    def get_double_lower_bounds(self):
+        """Returns an array of the lower bounds of the variables."""
+        values = []
+
+        for node, x in self.nodes.items():
+            if x.is_variable:
+                values.extend(x.get_lower_bounds())
+
+        for parameter, x in self.parameters.items():
+            if x.is_variable:
+                values.extend(x.get_lower_bounds())
+
+        if self.days_of_year is not None:
+            if self.days_of_year.is_variable:
+                values.extend(self.days_of_year.get_lower_bounds())
+
+        if self.y.is_variable:
+            values.extend(self.y.get_lower_bounds())
+
+        return np.array(values)
+
+    def value(self, ts, scenario_index):
+        """Calculate the interpolate Rbf value from the current state."""
+        # Use the cached node and parameter orders so that the exogenous inputs
+        # are in the correct order.
+        nodes = self._node_order
+        parameters = self._parameter_order
+        days_of_year = self.days_of_year
+
+        # Create the arguments for the Rbf function.
+        args = []
+        for node in nodes:
+            if isinstance(node, Storage):
+                # Storage nodes use the current volume
+                x = node.current_pc[scenario_index.global_id]
+            else:
+                # Other nodes are based on the flow
+                x = node.flow[scenario_index.global_id]
+            args.append(x)
+
+        for parameter in parameters:
+            x = parameter.get_value(scenario_index)
+            args.append(x)
+
+        if days_of_year is not None:
+            # Normalise DoY to be between 0 and 1.
+            x = 2 * np.pi * (ts.dayofyear - 1) / 365
+            args.append(np.sin(x))
+            args.append(np.cos(x))
+
+        # Perform interpolation.
+        return self._rbf_func(*args)
```

Comment:
Add a (optional?) clipping to make this return a value >= 0?

```
return max(self._rbf_func(*args), 0)
```

Reply:
@jetuk, Before adding the clipping I tried to run the optimisation but I got this:
```
    m = Model.load(self.pywr_model_json)
  File "pywr\_model.pyx", line 258, in pywr._model.Model.load
  File "pywr\_model.pyx", line 192, in pywr._model.Model.loads
  File "pywr\_model.pyx", line 303, in pywr._model.Model.load
  File "pywr\_model.pyx", line 86, in pywr._model.Model.__init__
  File "c:\users\n65847jg\dropbox (the university of manchester)\doctorado\pywr\pywr\solvers\__init__.py", line 156, in <module>
    from .cython_lpsolve import CythonLPSolveSolver as cy_CythonLPSolveSolver
  File "pywr\recorders\_recorders.pxd", line 6, in init pywr.solvers.cython_lpsolve
ValueError: pywr.recorders._recorders.Aggregator size changed, may indicate binary incompatibility. Expected 56 from C header, got 40 from PyObject
```
Additionally, I changed the `timestep` freq to "D" and the error is:
```
  File "c:\users\n65847jg\dropbox (the university of manchester)\doctorado\pywr\pywr\optimisation\__init__.py", line 102, in make_model
    m = Model.load(self.pywr_model_json)
  File "pywr\_model.pyx", line 258, in pywr._model.Model.load
  File "pywr\_model.pyx", line 192, in pywr._model.Model.loads
  File "pywr\_model.pyx", line 300, in pywr._model.Model.load
ValueError: invalid literal for int() with base 10: 'D'
``` 
it is supposed having merged `master` the pandas' freq may be available. isn't it?

Reply:
Did you recompile?

Reply:
Thanks. I had compiled but I did not `pull` the changes before the recompilation :flushed: